csharp/src/Google.Protobuf/JsonFormatter.cs

```diff
@@ -780,6 +783,62 @@ private static void HexEncodeUtf16CodeUnit(TextWriter writer, char c)
             writer.Write(Hex[(c >> 4) & 0xf]);
             writer.Write(Hex[(c >> 0) & 0xf]);
         }
+        
+        
+        private void WriteJsonObjectOpen(TextWriter writer, char openChar = '{')
+        {
+            writer.Write(openChar);
+            if (!settings.Pretty)
+            {
+                writer.Write(' ');
+            }
+        }
+
+        private void WriteJsonObjectClose(TextWriter writer, bool hasFields, int depth, char closeChar = '}')
+        {
+            if (hasFields)
+            {
+                if (settings.Pretty)
+                {
+                    writer.WriteLine();
+                    WriteIndentation(writer, depth);
+                }
+                else
+                {
+                    writer.Write(" ");
+                }
+            }
+            
+            writer.Write(closeChar);
+        }
+
+        private void WriteJsonObjectEntry(TextWriter writer, bool first, int depth)
```

Comment:
Writes a separator between entries in a list or a dictionary.

This depends on whether it's the first item, and for pretty printing the logic is tricky enough to not want to have it in multiple places.

The separators are written _before_ the entry, not after. Same as original implementation. I suppose the benefit is that empty objects / lists will be on a single line `{}`, `[]`.

```
{(
  entry 1)(,
  entry 2)(,
  entry 3)
}
```